genie-agent/src/main/java/com/netflix/genie/agent/jobexecution/statemachine/listeners/LoggingListener.java

```diff
@@ -0,0 +1,231 @@
+/*
+ *
+ *  Copyright 2018 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.agent.jobexecution.statemachine.listeners;
+
+import com.netflix.genie.agent.jobexecution.statemachine.Events;
+import com.netflix.genie.agent.jobexecution.statemachine.States;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.messaging.Message;
+import org.springframework.statemachine.StateContext;
+import org.springframework.statemachine.StateMachine;
+import org.springframework.statemachine.action.Action;
+import org.springframework.statemachine.state.State;
+import org.springframework.statemachine.transition.Transition;
+import org.springframework.statemachine.trigger.Trigger;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.Nullable;
+
+/**
+ * Listener that logs state machine events and transitions.
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+@Component
+@Lazy
+public class LoggingListener implements JobExecutionListener {
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void stateChanged(final State<States, Events> from, final State<States, Events> to) {
+        log.info(
```

Comment:
All of these should be debug in production. Either that or this entire configuration should be `ConditionalOnProperty` so you can turn it off and on. The second one might be better as then it would never even be created or wired in unless it was desired and save the processing of calling these methods.

Reply:
I think this is the most important piece of logging for debugging when someone's execution goes wrong.
We can ask them to provide the logs (or see them on server eventually) and know exactly what happened.

The number of states/transitions is relatively low (10) so this should not be too verbose, we may turn off some of them we don't really care for (such as transitions, since we don't have transition actions).

Reply:
This is fine provided we do not output these log messages to `stdout`. I know that for the runner in the past when we had a bunch of the genie output going to std out people/dan/etc were complaining about it. The Genie agent execution should be transparent. I'm fine with this going into a genie specific log every time though. Just not overdo `info` logging